@ahus1 Thanks for the changes. The `java.lang.RuntimeException: No org.keycloak.cluster.ManagedCacheManagerProvider found. If running in embedded mode set the [embedded] property to this provider.` error is away. Now we have the following error in 3 log files (0_Store Model Tests (7).txt, 0_Store Model Tests (16).txt, 0_Store Model Tests (28).txt)
```
ISPN000136: Error executing command PutMapCommand on Cache 'offlineClientSessions', writing keys [f815e1f8-af70-41f8-b303-27bd35c28e99, b61c1ef5-632a-440d-90d8-40c80f52f2ef, f659082c-61d3-440c-83ca-0d5eb488ef91, b4b4d8d2-2fb1-4072-ab27-a533de4e01b8, 234b4e7c-1a43-4acc-9137-170e4511e058, a5112574-20ae-4b88-b5cc-22434e9265f7, 3a07f5a6-360f-4e70-929e-409c2a2d0b3b, 9567600d-4cc9-4461-9b31-f35ea1abd84c...<12 other elements>]
2023-11-01T20:57:53.6057611Z org.infinispan.util.concurrent.TimeoutException: ISPN000427: Timeout after 15 seconds waiting for acks ([node-22(site-id=site-2), node-24(site-id=site-2)]). Id=100420
```

I think, I saw this several times before. Probably it's not cased by your changes. Still, we have 3 occurrences out of 35. Do you want to investigate it now or file a separate issue?

Full stacktrace :
```
ERROR [org.infinispan.interceptors.impl.InvocationContextInterceptor] (timeout-thread-node-24-p124-t1) ISPN000136: Error executing command PutMapCommand on Cache 'offlineClientSessions', writing keys [f815e1f8-af70-41f8-b303-27bd35c28e99, b61c1ef5-632a-440d-90d8-40c80f52f2ef, f659082c-61d3-440c-83ca-0d5eb488ef91, b4b4d8d2-2fb1-4072-ab27-a533de4e01b8, 234b4e7c-1a43-4acc-9137-170e4511e058, a5112574-20ae-4b88-b5cc-22434e9265f7, 3a07f5a6-360f-4e70-929e-409c2a2d0b3b, 9567600d-4cc9-4461-9b31-f35ea1abd84c...<12 other elements>]
org.infinispan.util.concurrent.TimeoutException: ISPN000427: Timeout after 15 seconds waiting for acks ([node-22(site-id=site-2), node-24(site-id=site-2)]). Id=100420
	at org.infinispan.util.concurrent.CommandAckCollector.createTimeoutException(CommandAckCollector.java:228)
	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:281)
	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:261)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
ERROR [org.infinispan.interceptors.impl.InvocationContextInterceptor] (timeout-thread-node-22-p134-t1) ISPN000136: Error executing command PutMapCommand on Cache 'offlineClientSessions', writing keys [cf176276-ff1c-4326-8c65-a7ec374c945e, f8c99fb0-e0a1-405d-bd42-cb57da735c63, c65fa9d8-32f2-4001-8e76-936f5450cc3f, f758597b-2e49-4586-ab3f-67ca64d5ae8e, 5b2aa161-b516-4d76-80c3-88fdf01b0a0d, 62fabd0c-4e88-492d-aa2f-4510d4a2ad4d, 3693f555-70df-4e06-8751-518e17a5d36f, f559e6da-0e6f-4915-aeab-e9511cab9de3...<25 other elements>]
org.infinispan.util.concurrent.TimeoutException: ISPN000427: Timeout after 15 seconds waiting for acks ([node-22(site-id=site-2)]). Id=100422
	at org.infinispan.util.concurrent.CommandAckCollector.createTimeoutException(CommandAckCollector.java:228)
	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:281)
	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:261)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
ERROR [org.infinispan.interceptors.impl.InvocationContextInterceptor] (timeout-thread-node-24-p124-t1) ISPN000136: Error executing command PutMapCommand on Cache 'offlineClientSessions', writing keys [92ba8484-020a-4928-99ab-6e249b10e4c3, 063a139b-9fff-4032-96df-6019d3c54d2e, 76530a5e-1fde-465f-a17d-d7fac887e7cd, 3e6dfcb2-1f91-44c0-9c52-d757e96a48e3, c38e2b15-09be-4464-9283-b3d447e1ea75, 28d76f85-c7e6-4431-a9b4-067140f8657b, 714cb7cc-5270-46d8-b1d7-323ab2eef56f, 94f06f30-e8ee-45c8-bcf9-f4489afcebae...<37 other elements>]
org.infinispan.util.concurrent.TimeoutException: ISPN000427: Timeout after 15 seconds waiting for acks ([node-22(site-id=site-2)]). Id=100421
 	at org.infinispan.util.concurrent.CommandAckCollector.createTimeoutException(CommandAckCollector.java:228)
 	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:281)
 	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:261)
 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
 	at java.base/java.lang.Thread.run(Thread.java:840)
WARN  [org.infinispan.CLUSTER] (jgroups-11,node-22(site-id=site-2)) ISPN000071: Caught exception when handling command org.infinispan.manager.impl.ReplicableManagerFunctionCommand@28ea04e1
org.infinispan.util.concurrent.TimeoutException: ISPN000427: Timeout after 15 seconds waiting for acks ([node-22(site-id=site-2)]). Id=100422
 	at org.infinispan.interceptors.impl.AsyncInterceptorChainImpl.invoke(AsyncInterceptorChainImpl.java:258)
 	at org.infinispan.cache.impl.InvocationHelper.doInvoke(InvocationHelper.java:323)
 	at org.infinispan.cache.impl.InvocationHelper.invoke(InvocationHelper.java:111)
 	at org.infinispan.cache.impl.InvocationHelper.invoke(InvocationHelper.java:93)
 	at org.infinispan.cache.impl.CacheImpl.putAll(CacheImpl.java:1361)
 	at org.infinispan.cache.impl.DecoratedCache.putAll(DecoratedCache.java:562)
 	at org.infinispan.cache.impl.DecoratedCache.putAll(DecoratedCache.java:567)
 	at org.infinispan.cache.impl.EncoderCache.putAll(EncoderCache.java:792)
 	at org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheSessionsLoader.loadSessions(RemoteCacheSessionsLoader.java:144)
 	at org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheSessionsLoader.loadSessions(RemoteCacheSessionsLoader.java:47)
 	at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker$1.run(SessionInitializerWorker.java:71)
 	at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:262)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:384)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:261)
 	at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker.apply(SessionInitializerWorker.java:67)
 	at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker.apply(SessionInitializerWorker.java:34)
 	at org.infinispan.manager.impl.ReplicableManagerFunctionCommand.invokeAsync(ReplicableManagerFunctionCommand.java:45)
 	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler$ReplicableCommandRunner.run(GlobalInboundInvocationHandler.java:256)
 	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleReplicableCommand(GlobalInboundInvocationHandler.java:175)
 	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleFromCluster(GlobalInboundInvocationHandler.java:114)
 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1548)
 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1476)
 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.lambda$up$1(JGroupsTransport.java:1695)
 	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1687)
 	at org.jgroups.JChannel.up(JChannel.java:749)
 	at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:939)
 	at org.jgroups.protocols.FRAG2.up(FRAG2.java:161)
 	at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
 	at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
 	at org.jgroups.protocols.pbcast.GMS.up(GMS.java:859)
 	at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:246)
 	at org.jgroups.protocols.UNICAST3.deliverBatch(UNICAST3.java:1108)
 	at org.jgroups.protocols.UNICAST3.handleBatchReceived(UNICAST3.java:882)
 	at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:549)
 	at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:722)
 	at org.jgroups.protocols.VERIFY_SUSPECT.up(VERIFY_SUSPECT.java:146)
 	at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:193)
 	at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:202)
 	at org.jgroups.protocols.MERGE3.up(MERGE3.java:288)
 	at org.jgroups.protocols.Discovery.up(Discovery.java:314)
 	at org.jgroups.protocols.RED.up(RED.java:119)
 	at org.jgroups.protocols.TP.passBatchUp(TP.java:1210)
 	at org.jgroups.util.SubmitToThreadPool$BatchHandler.passBatchUp(SubmitToThreadPool.java:161)
 	at org.jgroups.util.SubmitToThreadPool$BatchHandler.run(SubmitToThreadPool.java:157)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
 	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: org.infinispan.util.concurrent.TimeoutException: ISPN000427: Timeout after 15 seconds waiting for acks ([node-22(site-id=site-2)]). Id=100422
 	at org.infinispan.util.concurrent.CommandAckCollector.createTimeoutException(CommandAckCollector.java:228)
 	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:281)
 	at org.infinispan.util.concurrent.CommandAckCollector$BaseCollector.call(CommandAckCollector.java:261)
 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
 	... 3 more
 	Suppressed: org.infinispan.commons.util.logging.TraceException
 		at org.infinispan.interceptors.impl.SimpleAsyncInvocationStage.get(SimpleAsyncInvocationStage.java:39)
 		at org.infinispan.interceptors.impl.AsyncInterceptorChainImpl.invoke(AsyncInterceptorChainImpl.java:249)
 		at org.infinispan.cache.impl.InvocationHelper.doInvoke(InvocationHelper.java:323)
 		at org.infinispan.cache.impl.InvocationHelper.invoke(InvocationHelper.java:111)
 		at org.infinispan.cache.impl.InvocationHelper.invoke(InvocationHelper.java:93)
 		at org.infinispan.cache.impl.CacheImpl.putAll(CacheImpl.java:1361)
 		at org.infinispan.cache.impl.DecoratedCache.putAll(DecoratedCache.java:562)
 		at org.infinispan.cache.impl.DecoratedCache.putAll(DecoratedCache.java:567)
 		at org.infinispan.cache.impl.EncoderCache.putAll(EncoderCache.java:792)
 		at org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheSessionsLoader.loadSessions(RemoteCacheSessionsLoader.java:144)
 		at org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheSessionsLoader.loadSessions(RemoteCacheSessionsLoader.java:47)
 		at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker$1.run(SessionInitializerWorker.java:71)
 		at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:262)
 		at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:384)
 		at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:261)
 		at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker.apply(SessionInitializerWorker.java:67)
 		at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker.apply(SessionInitializerWorker.java:34)
 		at org.infinispan.manager.impl.ReplicableManagerFunctionCommand.invokeAsync(ReplicableManagerFunctionCommand.java:45)
 		at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler$ReplicableCommandRunner.run(GlobalInboundInvocationHandler.java:256)
 		at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleReplicableCommand(GlobalInboundInvocationHandler.java:175)
 		at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleFromCluster(GlobalInboundInvocationHandler.java:114)
 		at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1548)
 		at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1476)
 		at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.lambda$up$1(JGroupsTransport.java:1695)
 		at java.base/java.lang.Iterable.forEach(Iterable.java:75)
 		at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1687)
 		at org.jgroups.JChannel.up(JChannel.java:749)
 		at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:939)
 		at org.jgroups.protocols.FRAG2.up(FRAG2.java:161)
 		at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
 		at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
 		at org.jgroups.protocols.pbcast.GMS.up(GMS.java:859)
 		at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:246)
 		at org.jgroups.protocols.UNICAST3.deliverBatch(UNICAST3.java:1108)
 		at org.jgroups.protocols.UNICAST3.handleBatchReceived(UNICAST3.java:882)
 		at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:549)
 		at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:722)
 		at org.jgroups.protocols.VERIFY_SUSPECT.up(VERIFY_SUSPECT.java:146)
 		at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:193)
 		at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:202)
 		at org.jgroups.protocols.MERGE3.up(MERGE3.java:288)
 		at org.jgroups.protocols.Discovery.up(Discovery.java:314)
 		at org.jgroups.protocols.RED.up(RED.java:119)
 		at org.jgroups.protocols.TP.passBatchUp(TP.java:1210)
 		at org.jgroups.util.SubmitToThreadPool$BatchHandler.passBatchUp(SubmitToThreadPool.java:161)
 		at org.jgroups.util.SubmitToThreadPool$BatchHandler.run(SubmitToThreadPool.java:157)
 		... 3 more
ERROR [org.keycloak.models.sessions.infinispan.initializer.InfinispanCacheInitializer] (pool-40-thread-3) Interruped exception when computed future. Errors: 1
 java.lang.InterruptedException
 	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:386)
 	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
 	at org.keycloak.models.sessions.infinispan.initializer.InfinispanCacheInitializer.startLoadingImpl(InfinispanCacheInitializer.java:183)
 	at org.keycloak.models.sessions.infinispan.initializer.InfinispanCacheInitializer.startLoading(InfinispanCacheInitializer.java:113)
 	at org.keycloak.models.sessions.infinispan.initializer.CacheInitializer.loadSessions(CacheInitializer.java:66)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory$6.run(InfinispanUserSessionProviderFactory.java:343)
 	at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:262)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:384)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:261)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory.loadSessionsFromRemoteCache(InfinispanUserSessionProviderFactory.java:329)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory.loadSessionsFromRemoteCaches(InfinispanUserSessionProviderFactory.java:321)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory.access$800(InfinispanUserSessionProviderFactory.java:72)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory$1.lambda$onEvent$0(InfinispanUserSessionProviderFactory.java:129)
 	at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:262)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:384)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:261)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithTimeout(KeycloakModelUtils.java:475)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory$1.onEvent(InfinispanUserSessionProviderFactory.java:123)
 	at org.keycloak.services.DefaultKeycloakSessionFactory.publish(DefaultKeycloakSessionFactory.java:91)
 	at org.keycloak.testsuite.model.KeycloakModelTest.createKeycloakSessionFactory(KeycloakModelTest.java:349)
 	at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:496)
 	at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$6(KeycloakModelTest.java:401)
 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
 	at java.base/java.lang.Thread.run(Thread.java:840)
DEBUG [org.keycloak.testsuite.model.KeycloakModelParameters] (pool-40-thread-4) Closing KeycloakSessionFactory 25 (from pool-40-thread-4 thread)
ERROR [org.keycloak.models.sessions.infinispan.initializer.CacheInitializer] (pool-40-thread-1) Interrupted
 java.lang.InterruptedException: sleep interrupted
 	at java.base/java.lang.Thread.sleep(Native Method)
 	at java.base/java.lang.Thread.sleep(Thread.java:344)
 	at java.base/java.util.concurrent.TimeUnit.sleep(TimeUnit.java:446)
 	at org.keycloak.models.sessions.infinispan.initializer.CacheInitializer.loadSessions(CacheInitializer.java:41)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory$6.run(InfinispanUserSessionProviderFactory.java:343)
 	at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:262)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:384)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:261)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory.loadSessionsFromRemoteCache(InfinispanUserSessionProviderFactory.java:329)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory.loadSessionsFromRemoteCaches(InfinispanUserSessionProviderFactory.java:321)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory.access$800(InfinispanUserSessionProviderFactory.java:72)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory$1.lambda$onEvent$0(InfinispanUserSessionProviderFactory.java:129)
 	at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:262)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:384)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:261)
 	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithTimeout(KeycloakModelUtils.java:475)
 	at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory$1.onEvent(InfinispanUserSessionProviderFactory.java:123)
 	at org.keycloak.services.DefaultKeycloakSessionFactory.publish(DefaultKeycloakSessionFactory.java:91)
 	at org.keycloak.testsuite.model.KeycloakModelTest.createKeycloakSessionFactory(KeycloakModelTest.java:349)
 	at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:496)
 	at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$6(KeycloakModelTest.java:401)
 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
 	at java.base/java.lang.Thread.run(Thread.java:840)
```

--- a/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/InfinispanClusterProviderFactory.java
+++ b/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/InfinispanClusterProviderFactory.java
@@ -19,6 +19,7 @@ package org.keycloak.cluster.infinispan;
 
 import org.infinispan.Cache;
 import org.infinispan.client.hotrod.exceptions.HotRodClientException;
+import org.infinispan.lifecycle.ComponentStatus;
 import org.infinispan.manager.EmbeddedCacheManager;
 import org.infinispan.notifications.Listener;
 import org.infinispan.notifications.cachemanagerlistener.annotation.ViewChanged;
@@ -230,7 +231,11 @@ public class InfinispanClusterProviderFactory implements ClusterProviderFactory,
                             Tracked in https://github.com/keycloak/keycloak/issues/9871
                         */
                         synchronized (DefaultInfinispanConnectionProviderFactory.class) {
-                            workCache.entrySet().removeIf(new LockEntryPredicate(removedNodesAddresses));
+                            if (workCache.getStatus() == ComponentStatus.RUNNING) {
+                                workCache.entrySet().removeIf(new LockEntryPredicate(removedNodesAddresses));
+                            } else {
+                                logger.warn("work cache is not running, ignoring event");
+                            }
                         }
                     }
                 } catch (Throwable t) {

--- a/model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java
+++ b/model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java
@@ -66,17 +66,17 @@ import static org.keycloak.models.cache.infinispan.InfinispanCacheRealmProviderF
  */
 public class DefaultInfinispanConnectionProviderFactory implements InfinispanConnectionProviderFactory, EnvironmentDependentProviderFactory {
 
-    protected static final Logger logger = Logger.getLogger(DefaultInfinispanConnectionProviderFactory.class);
+    private static final Logger logger = Logger.getLogger(DefaultInfinispanConnectionProviderFactory.class);
 
-    protected Config.Scope config;
+    private Config.Scope config;
 
-    protected EmbeddedCacheManager cacheManager;
+    private volatile EmbeddedCacheManager cacheManager;
 
-    protected RemoteCacheProvider remoteCacheProvider;
+    private volatile RemoteCacheProvider remoteCacheProvider;
 
-    protected boolean containerManaged;
+    protected volatile boolean containerManaged;
 
-    private TopologyInfo topologyInfo;
+    private volatile TopologyInfo topologyInfo;
 
     @Override
     public InfinispanConnectionProvider create(KeycloakSession session) {
@@ -102,7 +102,6 @@ public class DefaultInfinispanConnectionProviderFactory implements InfinispanCon
             if (remoteCacheProvider != null) {
                 remoteCacheProvider.stop();
             }
-            cacheManager = null;
         }
     }
 
@@ -142,62 +141,67 @@ public class DefaultInfinispanConnectionProviderFactory implements InfinispanCon
                         
                         managedCacheManager = provider.getCacheManager(config);
                     }
-                    
+
+                    // store it in a locale variable first, so it is not visible to the outside, yet
+                    EmbeddedCacheManager localCacheManager;
                     if (managedCacheManager == null) {
                         if (!config.getBoolean("embedded", false)) {
                             throw new RuntimeException("No " + ManagedCacheManagerProvider.class.getName() + " found. If running in embedded mode set the [embedded] property to this provider.");
                         }
-                        initEmbedded();
+                        localCacheManager = initEmbedded();
                     } else {
-                        initContainerManaged(managedCacheManager);
+                        localCacheManager = initContainerManaged(managedCacheManager);
                     }
 
                     logger.infof(topologyInfo.toString());
 
-                    remoteCacheProvider = new RemoteCacheProvider(config, cacheManager);
+                    remoteCacheProvider = new RemoteCacheProvider(config, localCacheManager);
+                    // only set the cache manager attribute at the very end to avoid passing a half-initialized entry callers
+                    cacheManager = localCacheManager;
                 }
             }
         }
     }
 
-    protected void initContainerManaged(EmbeddedCacheManager cacheManager) {
-        this.cacheManager = cacheManager;
+    protected EmbeddedCacheManager initContainerManaged(EmbeddedCacheManager cacheManager) {
         containerManaged = true;
 
-        long realmRevisionsMaxEntries = this.cacheManager.getCache(InfinispanConnectionProvider.REALM_CACHE_NAME).getCacheConfiguration().memory().size();
+        long realmRevisionsMaxEntries = cacheManager.getCache(InfinispanConnectionProvider.REALM_CACHE_NAME).getCacheConfiguration().memory().size();
         realmRevisionsMaxEntries = realmRevisionsMaxEntries > 0
                 ? 2 * realmRevisionsMaxEntries
                 : InfinispanConnectionProvider.REALM_REVISIONS_CACHE_DEFAULT_MAX;
 
-        this.cacheManager.defineConfiguration(InfinispanConnectionProvider.REALM_REVISIONS_CACHE_NAME, getRevisionCacheConfig(realmRevisionsMaxEntries));
-        this.cacheManager.getCache(InfinispanConnectionProvider.REALM_REVISIONS_CACHE_NAME, true);
+        cacheManager.defineConfiguration(InfinispanConnectionProvider.REALM_REVISIONS_CACHE_NAME, getRevisionCacheConfig(realmRevisionsMaxEntries));
+        cacheManager.getCache(InfinispanConnectionProvider.REALM_REVISIONS_CACHE_NAME, true);
 
-        long userRevisionsMaxEntries = this.cacheManager.getCache(InfinispanConnectionProvider.USER_CACHE_NAME).getCacheConfiguration().memory().size();
+        long userRevisionsMaxEntries = cacheManager.getCache(InfinispanConnectionProvider.USER_CACHE_NAME).getCacheConfiguration().memory().size();
         userRevisionsMaxEntries = userRevisionsMaxEntries > 0
                 ? 2 * userRevisionsMaxEntries
                 : InfinispanConnectionProvider.USER_REVISIONS_CACHE_DEFAULT_MAX;
 
-        this.cacheManager.defineConfiguration(InfinispanConnectionProvider.USER_REVISIONS_CACHE_NAME, getRevisionCacheConfig(userRevisionsMaxEntries));
-        this.cacheManager.getCache(InfinispanConnectionProvider.USER_REVISIONS_CACHE_NAME, true);
-        this.cacheManager.getCache(InfinispanConnectionProvider.AUTHORIZATION_CACHE_NAME, true);
-        this.cacheManager.getCache(InfinispanConnectionProvider.AUTHENTICATION_SESSIONS_CACHE_NAME, true);
-        this.cacheManager.getCache(InfinispanConnectionProvider.KEYS_CACHE_NAME, true);
-        this.cacheManager.getCache(InfinispanConnectionProvider.ACTION_TOKEN_CACHE, true);
+        cacheManager.defineConfiguration(InfinispanConnectionProvider.USER_REVISIONS_CACHE_NAME, getRevisionCacheConfig(userRevisionsMaxEntries));
+        cacheManager.getCache(InfinispanConnectionProvider.USER_REVISIONS_CACHE_NAME, true);
+        cacheManager.getCache(InfinispanConnectionProvider.AUTHORIZATION_CACHE_NAME, true);
+        cacheManager.getCache(InfinispanConnectionProvider.AUTHENTICATION_SESSIONS_CACHE_NAME, true);
+        cacheManager.getCache(InfinispanConnectionProvider.KEYS_CACHE_NAME, true);
+        cacheManager.getCache(InfinispanConnectionProvider.ACTION_TOKEN_CACHE, true);
 
-        long authzRevisionsMaxEntries = this.cacheManager.getCache(InfinispanConnectionProvider.AUTHORIZATION_CACHE_NAME).getCacheConfiguration().memory().size();
+        long authzRevisionsMaxEntries = cacheManager.getCache(InfinispanConnectionProvider.AUTHORIZATION_CACHE_NAME).getCacheConfiguration().memory().size();
         authzRevisionsMaxEntries = authzRevisionsMaxEntries > 0
                 ? 2 * authzRevisionsMaxEntries
                 : InfinispanConnectionProvider.AUTHORIZATION_REVISIONS_CACHE_DEFAULT_MAX;
 
-        this.cacheManager.defineConfiguration(InfinispanConnectionProvider.AUTHORIZATION_REVISIONS_CACHE_NAME, getRevisionCacheConfig(authzRevisionsMaxEntries));
-        this.cacheManager.getCache(InfinispanConnectionProvider.AUTHORIZATION_REVISIONS_CACHE_NAME, true);
+        cacheManager.defineConfiguration(InfinispanConnectionProvider.AUTHORIZATION_REVISIONS_CACHE_NAME, getRevisionCacheConfig(authzRevisionsMaxEntries));
+        cacheManager.getCache(InfinispanConnectionProvider.AUTHORIZATION_REVISIONS_CACHE_NAME, true);
 
-        this.topologyInfo = new TopologyInfo(this.cacheManager, config, false);
+        this.topologyInfo = new TopologyInfo(cacheManager, config, false);
 
         logger.debugv("Using container managed Infinispan cache container, lookup={0}", cacheManager);
+
+        return cacheManager;
     }
 
-    protected void initEmbedded() {
+    protected EmbeddedCacheManager initEmbedded() {
         GlobalConfigurationBuilder gcb = new GlobalConfigurationBuilder();
 
         boolean clustered = config.getBoolean("clustered", false);
@@ -221,7 +225,7 @@ public class DefaultInfinispanConnectionProviderFactory implements InfinispanCon
         // See https://infinispan.org/docs/stable/titles/developing/developing.html#marshalling for the details
         gcb.serialization().marshaller(new JBossUserMarshaller());
 
-        cacheManager = new DefaultCacheManager(gcb.build());
+        EmbeddedCacheManager cacheManager = new DefaultCacheManager(gcb.build());
         if (useKeycloakTimeService) {
             setTimeServiceToKeycloakTime(cacheManager);
         }
@@ -374,6 +378,8 @@ public class DefaultInfinispanConnectionProviderFactory implements InfinispanCon
 
         cacheManager.defineConfiguration(InfinispanConnectionProvider.AUTHORIZATION_REVISIONS_CACHE_NAME, getRevisionCacheConfig(authzRevisionsMaxEntries));
         cacheManager.getCache(InfinispanConnectionProvider.AUTHORIZATION_REVISIONS_CACHE_NAME, true);
+
+        return cacheManager;
     }
 
     private Configuration getRevisionCacheConfig(long maxEntries) {

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/LegacyInfinispanConnectionFactory.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/LegacyInfinispanConnectionFactory.java
@@ -29,11 +29,12 @@ public class LegacyInfinispanConnectionFactory extends DefaultInfinispanConnecti
         implements EnvironmentDependentProviderFactory {
 
     @Override
-    protected void initContainerManaged(EmbeddedCacheManager cacheManager) {
-        super.initContainerManaged(cacheManager);
+    protected EmbeddedCacheManager initContainerManaged(EmbeddedCacheManager cacheManager) {
+        EmbeddedCacheManager result = super.initContainerManaged(cacheManager);
         // force closing the cache manager when stopping the provider
         // we probably want to refactor the default impl a bit to support this use case
         containerManaged = false;
+        return result;
     }
 
     @Override

--- a/testsuite/model/pom.xml
+++ b/testsuite/model/pom.xml
@@ -173,6 +173,8 @@
                     <!-- See also https://github.com/wildfly/wildfly-core/blob/7e5624cf92ebe4b64a4793a8c0b2a340c0d6d363/core-feature-pack/common/src/main/resources/content/bin/common.sh#L57-L60 -->
                     <argLine>@{argLine} -Xmx1536m -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED</argLine>
                     <systemPropertyVariables>
+                        <!-- ensure a log entry if we're running out of JGroup threads -->
+                        <jgroups.thread_dumps_threshold>1</jgroups.thread_dumps_threshold>
                         <!-- keycloak.model.parameters lists parameter classes from 
                              org.keycloak.model.parameters package and determine enabled providers -->
                         <keycloak.model.parameters>${keycloak.model.parameters}</keycloak.model.parameters>

--- a/testsuite/model/src/main/java/org/keycloak/testsuite/model/Config.java
+++ b/testsuite/model/src/main/java/org/keycloak/testsuite/model/Config.java
@@ -24,6 +24,7 @@ import org.keycloak.common.util.SystemEnvProperties;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.BooleanSupplier;
 import java.util.stream.Collectors;
 
@@ -35,7 +36,7 @@ public class Config implements ConfigProvider {
 
     private final Properties systemProperties = new SystemEnvProperties();
 
-    private final Map<String, String> defaultProperties = new HashMap<>();
+    private final Map<String, String> defaultProperties = new ConcurrentHashMap<>();
     private final ThreadLocal<Map<String, String>> properties = new ThreadLocal<Map<String, String>>() {
         @Override
         protected Map<String, String> initialValue() {

